File: server/router/flashcard.ts
import { z } from 'zod';
import { router, publicProcedure } from './trpc.js';
import { db, schema } from '../db/index.js';
import { eq, lte, desc, like } from 'drizzle-orm';
import { ulid } from 'ulid';
import { supermemo, SuperMemoItem, SuperMemoGrade } from 'supermemo';
import { generateFlashcardsFromContent, generateFlashcardsForTopic } from '../ai/flashcards.js';

export const flashcardRouter = router({
  // List all flashcards
  list: publicProcedure
    .input(
      z.object({
        limit: z.number().min(1).max(100).default(50),
        offset: z.number().min(0).default(0),
      }).optional()
    )
    .query(async ({ input }) => {
      const { limit = 50, offset = 0 } = input || {};
      return db
        .select()
        .from(schema.flashcards)
        .orderBy(desc(schema.flashcards.createdAt))
        .limit(limit)
        .offset(offset);
    }),

  // Get cards due for review
  getDue: publicProcedure.query(async () => {
    const now = new Date();
    return db
      .select()
      .from(schema.flashcards)
      .where(lte(schema.flashcards.nextReview, now))
      .orderBy(schema.flashcards.nextReview)
      .limit(20);
  }),

  // Get a single flashcard
  get: publicProcedure
    .input(z.object({ id: z.string() }))
    .query(async ({ input }) => {
      const [card] = await db
        .select()
        .from(schema.flashcards)
        .where(eq(schema.flashcards.id, input.id))
        .limit(1);
      return card || null;
    }),

  // Create a new flashcard
  create: publicProcedure
    .input(
      z.object({
        front: z.string(),
        back: z.string(),
        sourceId: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      const now = new Date();
      const id = ulid();

      await db.insert(schema.flashcards).values({
        id,
        front: input.front,
        back: input.back,
        sourceId: input.sourceId,
        easiness: 2.5,
        interval: 0,
        repetitions: 0,
        nextReview: now,
        createdAt: now,
      });

      return { id };
    }),

  // Review a flashcard (SM-2 algorithm)
  review: publicProcedure
    .input(
      z.object({
        id: z.string(),
        grade: z.number().min(0).max(5), // SM-2 grade: 0-5
      })
    )
    .mutation(async ({ input }) => {
      const [card] = await db
        .select()
        .from(schema.flashcards)
        .where(eq(schema.flashcards.id, input.id))
        .limit(1);

      if (!card) throw new Error('Flashcard not found');

      const item: SuperMemoItem = {
        interval: card.interval || 0,
        repetition: card.repetitions || 0,
        efactor: card.easiness || 2.5,
      };

      const result = supermemo(item, input.grade as SuperMemoGrade);

      const nextReview = new Date();
      nextReview.setDate(nextReview.getDate() + result.interval);

      await db
        .update(schema.flashcards)
        .set({
          easiness: result.efactor,
          interval: result.interval,
          repetitions: result.repetition,
          nextReview,
        })
        .where(eq(schema.flashcards.id, input.id));

      return { nextReview, interval: result.interval };
    }),

  // Update a flashcard
  update: publicProcedure
    .input(
      z.object({
        id: z.string(),
        front: z.string().optional(),
        back: z.string().optional(),
      })
    )
    .mutation(async ({ input }) => {
      const { id, ...updates } = input;
      const values: Record<string, unknown> = {};

      if (updates.front !== undefined) values.front = updates.front;
      if (updates.back !== undefined) values.back = updates.back;

      if (Object.keys(values).length > 0) {
        await db.update(schema.flashcards).set(values).where(eq(schema.flashcards.id, id));
      }
    }),

  // Delete a flashcard
  delete: publicProcedure
    .input(z.object({ id: z.string() }))
    .mutation(async ({ input }) => {
      await db.delete(schema.flashcards).where(eq(schema.flashcards.id, input.id));
    }),

  // Generate flashcards from a journal entry
  generateFromJournal: publicProcedure
    .input(z.object({ journalId: z.string() }))
    .mutation(async ({ input }) => {
      // Get the journal entry
      const [journal] = await db
        .select()
        .from(schema.journalEntries)
        .where(eq(schema.journalEntries.id, input.journalId))
        .limit(1);

      if (!journal) throw new Error('Journal entry not found');

      // Generate flashcards using AI
      const generated = await generateFlashcardsFromContent(
        journal.content,
        journal.title || undefined
      );

      // Save the flashcards
      const now = new Date();
      const ids: string[] = [];

      for (const card of generated) {
        const id = ulid();
        ids.push(id);

        await db.insert(schema.flashcards).values({
          id,
          front: card.front,
          back: card.back,
          sourceId: input.journalId,
          easiness: 2.5,
          interval: 0,
          repetitions: 0,
          nextReview: now,
          createdAt: now,
        });
      }

      return { count: ids.length, ids };
    }),

  // Generate flashcards for a topic (for "quiz me" command)
  generateForTopic: publicProcedure
    .input(z.object({ topic: z.string() }))
    .mutation(async ({ input }) => {
      // Search for related journal entries to provide context
      const relatedEntries = await db
        .select()
        .from(schema.journalEntries)
        .where(like(schema.journalEntries.content, `%${input.topic}%`))
        .limit(3);

      const context = relatedEntries.length > 0
        ? relatedEntries.map((e) => e.content).join('\n\n---\n\n')
        : undefined;

      // Generate flashcards using AI
      const generated = await generateFlashcardsForTopic(input.topic, context);

      // Save the flashcards
      const now = new Date();
      const ids: string[] = [];

      for (const card of generated) {
        const id = ulid();
        ids.push(id);

        await db.insert(schema.flashcards).values({
          id,
          front: card.front,
          back: card.back,
          sourceId: null,
          easiness: 2.5,
          interval: 0,
          repetitions: 0,
          nextReview: now,
          createdAt: now,
        });
      }

      return { count: ids.length, ids };
    }),

  // Get flashcards by source (journal entry)
  getBySource: publicProcedure
    .input(z.object({ sourceId: z.string() }))
    .query(async ({ input }) => {
      return db
        .select()
        .from(schema.flashcards)
        .where(eq(schema.flashcards.sourceId, input.sourceId))
        .orderBy(desc(schema.flashcards.createdAt));
    }),

  // Get stats
  getStats: publicProcedure.query(async () => {
    const now = new Date();
    const allCards = await db.select().from(schema.flashcards);

    const total = allCards.length;
    const due = allCards.filter((c) => c.nextReview <= now).length;
    const mastered = allCards.filter((c) => (c.interval || 0) >= 21).length; // 21+ days interval = mastered
    const learning = total - mastered;

    return { total, due, mastered, learning };
  }),
});

File: backend/routes/flashcard.py
import json
import time
from datetime import datetime, timezone, timedelta
from flask import Blueprint, jsonify, request
from ulid import ULID
from backend.db.connection import get_db, row_to_dict
from backend.tags import tags_json

bp = Blueprint('flashcard', __name__, url_prefix='/api/flashcards')

# Tags are stored lowercased so "JavaScript" and "javascript" are one deck.
_TAG_FILTER_SQL = 'EXISTS (SELECT 1 FROM json_each(flashcards.tags) WHERE json_each.value = ?)'

_INSERT_CARD_SQL = (
    'INSERT INTO flashcards(id, front, back, tags, source_id, easiness, interval, repetitions, next_review, created_at)'
    ' VALUES (?,?,?,?,?,?,?,?,?,?)'
)


def _card_to_dict(row) -> dict:
    d = row_to_dict(row)
    d['tags'] = json.loads(d['tags']) if d.get('tags') else []
    return d


def _tag_clause() -> tuple[str, list[str]]:
    """(' AND <filter>', [tag]) for the optional ?tag= query param, or ('', [])."""
    tag = (request.args.get('tag') or '').strip().lower()
    if not tag:
        return '', []
    return f' AND {_TAG_FILTER_SQL}', [tag]


def _sm2(interval: int, repetitions: int, efactor: float, grade: int) -> tuple[int, int, float]:
    if grade >= 3:
        if repetitions == 0:
            new_interval = 1
        elif repetitions == 1:
            new_interval = 6
        else:
            new_interval = round(interval * efactor)
        new_reps = repetitions + 1
    else:
        new_interval = 1
        new_reps = 0
    new_ef = max(1.3, efactor + 0.1 - (5 - grade) * (0.08 + (5 - grade) * 0.02))
    return new_interval, new_reps, new_ef


@bp.get('')
def list_cards():
    limit = min(int(request.args.get('limit', 50)), 100)
    offset = int(request.args.get('offset', 0))
    clause, params = _tag_clause()
    rows = get_db().execute(
        f'SELECT * FROM flashcards WHERE 1=1{clause} ORDER BY created_at DESC LIMIT ? OFFSET ?',
        [*params, limit, offset],
    ).fetchall()
    return jsonify([_card_to_dict(r) for r in rows])


@bp.get('/due')
def get_due():
    now = int(time.time())
    clause, params = _tag_clause()
    rows = get_db().execute(
        f'SELECT * FROM flashcards WHERE next_review <= ?{clause} ORDER BY next_review LIMIT 20',
        [now, *params],
    ).fetchall()
    return jsonify([_card_to_dict(r) for r in rows])


@bp.get('/stats')
def get_stats():
    now = int(time.time())
    clause, params = _tag_clause()
    row = get_db().execute(
        f'SELECT COUNT(*) AS total,'
        f' COALESCE(SUM(next_review <= ?), 0) AS due,'
        f' COALESCE(SUM(interval >= 21), 0) AS mastered'
        f' FROM flashcards WHERE 1=1{clause}',
        [now, *params],
    ).fetchone()
    return jsonify({
        'total': row['total'], 'due': row['due'], 'mastered': row['mastered'],
        'learning': row['total'] - row['mastered'],
    })


@bp.get('/tags')
def list_tags():
    rows = get_db().execute(
        'SELECT je.value AS name, COUNT(*) AS count'
        ' FROM flashcards f JOIN json_each(f.tags) je'
        ' GROUP BY je.value ORDER BY count DESC, name'
    ).fetchall()
    return jsonify([{'name': r['name'], 'count': r['count']} for r in rows])


@bp.get('/by-source/<source_id>')
def get_by_source(source_id):
    rows = get_db().execute(
        'SELECT * FROM flashcards WHERE source_id=? ORDER BY created_at DESC', (source_id,)
    ).fetchall()
    return jsonify([_card_to_dict(r) for r in rows])


@bp.get('/<id>')
def get_card(id):
    row = get_db().execute('SELECT * FROM flashcards WHERE id=?', (id,)).fetchone()
    if not row:
        return jsonify({'error': 'Not found'}), 404
    return jsonify(_card_to_dict(row))


@bp.post('')
def create_card():
    body = request.json or {}
    if not body.get('front') or not body.get('back'):
        return jsonify({'error': 'front and back required'}), 400
    now = int(time.time())
    id = str(ULID())
    get_db().execute(
        _INSERT_CARD_SQL,
        (id, body['front'], body['back'], tags_json(body.get('tags')), body.get('sourceId'), 2.5, 0, 0, now, now),
    )
    get_db().commit()
    return jsonify({'id': id}), 201


@bp.post('/<id>/review')
def review_card(id):
    body = request.json or {}
    grade = int(body.get('grade', 0))
    db = get_db()
    row = db.execute('SELECT * FROM flashcards WHERE id=?', (id,)).fetchone()
    if not row:
        return jsonify({'error': 'Not found'}), 404
    new_interval, new_reps, new_ef = _sm2(
        row['interval'] or 0, row['repetitions'] or 0, row['easiness'] or 2.5, grade
    )
    next_review = int((datetime.now(timezone.utc) + timedelta(days=new_interval)).timestamp())
    db.execute(
        'UPDATE flashcards SET easiness=?, interval=?, repetitions=?, next_review=? WHERE id=?',
        (new_ef, new_interval, new_reps, next_review, id),
    )
    db.commit()
    from datetime import datetime as dt
    return jsonify({
        'nextReview': dt.fromtimestamp(next_review, tz=timezone.utc).isoformat(),
        'interval': new_interval,
    })


@bp.patch('/<id>')
def update_card(id):
    body = request.json or {}
    updates: dict = {}
    if 'front' in body:
        updates['front'] = body['front']
    if 'back' in body:
        updates['back'] = body['back']
    if 'tags' in body:
        updates['tags'] = tags_json(body['tags'])
    if not updates:
        return jsonify({'success': True})
    set_clause = ', '.join(f'{k}=?' for k in updates)
    get_db().execute(f'UPDATE flashcards SET {set_clause} WHERE id=?', [*updates.values(), id])
    get_db().commit()
    return jsonify({'success': True})


@bp.delete('/<id>')
def delete_card(id):
    get_db().execute('DELETE FROM flashcards WHERE id=?', (id,))
    get_db().commit()
    return jsonify({'success': True})


@bp.post('/generate-from-journal')
def generate_from_journal():
    from backend.ai.flashcards import generate_flashcards_from_content
    body = request.json or {}
    journal_id = body.get('journalId')
    if not journal_id:
        return jsonify({'error': 'journalId required'}), 400
    db = get_db()
    row = db.execute('SELECT * FROM journal_entries WHERE id=?', (journal_id,)).fetchone()
    if not row:
        return jsonify({'error': 'Journal entry not found'}), 404
    cards = generate_flashcards_from_content(row['content'], row['title'])
    tags = tags_json(body.get('tags'))
    now = int(time.time())
    ids = []
    for card in cards:
        id = str(ULID())
        ids.append(id)
        db.execute(
            _INSERT_CARD_SQL,
            (id, card['front'], card['back'], tags, journal_id, 2.5, 0, 0, now, now),
        )
    db.commit()
    return jsonify({'count': len(ids), 'ids': ids})


@bp.post('/generate-for-topic')
def generate_for_topic():
    from backend.ai.flashcards import generate_flashcards_for_topic
    body = request.json or {}
    topic = body.get('topic', '').strip()
    if not topic:
        return jsonify({'error': 'topic required'}), 400
    db = get_db()
    related = db.execute(
        "SELECT content FROM journal_entries WHERE content LIKE ? LIMIT 3",
        (f'%{topic}%',),
    ).fetchall()
    context = '\n\n---\n\n'.join(r['content'] for r in related) if related else None
    cards = generate_flashcards_for_topic(topic, context)
    tags = tags_json(body.get('tags'))
    now = int(time.time())
    ids = []
    for card in cards:
        id = str(ULID())
        ids.append(id)
        db.execute(
            _INSERT_CARD_SQL,
            (id, card['front'], card['back'], tags, None, 2.5, 0, 0, now, now),
        )
    db.commit()
    return jsonify({'count': len(ids), 'ids': ids})
